File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/ddd/shared/Sku.java
package pe.joedayz.microservicios.modulo01.ddd.shared;

import java.util.Objects;

/**
 * Value Object: SKU (Stock Keeping Unit), el codigo unico de un producto.
 *
 * <p>En el contexto de Catalogo/Inventario un producto se identifica por su SKU.
 * Se auto-valida el formato para que nunca exista un SKU malformado en el sistema.
 */
public record Sku(String value) {

    public Sku {
        Objects.requireNonNull(value, "sku no puede ser null");
        if (!value.matches("[A-Z0-9\\-]{3,20}")) {
            throw new IllegalArgumentException(
                    "SKU invalido: '" + value + "'. Debe ser 3-20 caracteres A-Z, 0-9 o '-'");
        }
    }

    public static Sku of(String value) {
        return new Sku(value);
    }

    @Override
    public String toString() {
        return value;
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Ddd/Shared/Sku.cs
using System.Text.RegularExpressions;

namespace JoeDayz.Microservicios.Modulo01.Ddd.Shared;

/// <summary>
/// Codigo de producto (Stock Keeping Unit) con formato validado.
/// Usa un regex generado en tiempo de compilacion (source generator de .NET),
/// asi no hay coste de interpretacion en runtime ni reflexion en AOT.
/// </summary>
public sealed partial record Sku
{
    public Sku(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value, nameof(value));
        if (!SkuPattern().IsMatch(value))
        {
            throw new ArgumentException(
                $"SKU invalido: '{value}'. Debe ser 3-20 caracteres A-Z, 0-9 o '-'", nameof(value));
        }

        Value = value;
    }

    public string Value { get; }

    public static Sku Of(string value) => new(value);

    public override string ToString() => Value;

    [GeneratedRegex("^[A-Z0-9\\-]{3,20}$")]
    private static partial Regex SkuPattern();
}
